src/main/java/net/fabricmc/installer/UnattendedMain.java

```diff
@@ -0,0 +1,182 @@
+
+        URLClassLoader launchClassLoader = new URLClassLoader(new URL[]{ launchData.launchJar.toUri().toURL() });
+
+        // Use method handle to keep the stacktrace clean
+        MethodHandle handle = MethodHandles.publicLookup().findStatic(launchClassLoader.loadClass(launchData.mainClass), "main", MethodType.methodType(void.class, String[].class));
+        handle.invokeExact(args);
+    }
+
+    // Validates and downloads/installs the server if required
+    private static LaunchData initialise() throws IOException {
+        Properties properties = readProperties();
+
+        String loaderVersion = Objects.requireNonNull(properties.getProperty("fabric-loader-version"), "no loader-version specified in " + INSTALL_CONFIG_NAME);
+        String gameVersion = Objects.requireNonNull(properties.getProperty("game-version"), "no game-version specified in " + INSTALL_CONFIG_NAME);
+
+        // 0.12 or higher is required
+        validateLoaderVersion(loaderVersion);
+
+        // Vanilla server jar
+        Path serverJar = SERVER_DIR.resolve(String.format("%s-server.jar", gameVersion));
+        // Includes the mc version as this jar contains intermediary
+        Path serverLaunchJar = SERVER_DIR.resolve(String.format("fabric-loader-server-%s-minecraft-%s", loaderVersion, gameVersion));
+
+        String mainClass = null;
+
+        if (Files.exists(serverJar) && Files.exists(serverLaunchJar)) {
+            try {
+                mainClass = readMainClass(serverLaunchJar);
+            } catch (IOException e) {
+                // Wont throw here, will try to reinstall
+                System.err.println("Failed to read main class from server launch jar: " + e.getMessage());
+            }
+        }
+
+        if (mainClass != null) {
+            // All seems good, no need to reinstall
+            return new LaunchData(serverJar, serverLaunchJar, mainClass);
+        }
+
+        // TODO improve log output, its quite messy atm
+        Files.createDirectories(SERVER_DIR);
+        ServerInstaller.install(SERVER_DIR, loaderVersion, gameVersion, InstallerProgress.CONSOLE, serverLaunchJar);
+
+        InstallerProgress.CONSOLE.updateProgress(Utils.BUNDLE.getString("progress.download.minecraft"));
+        MinecraftServerDownloader downloader = new MinecraftServerDownloader(gameVersion);
+        downloader.downloadMinecraftServer(serverJar);
+
+        mainClass = readMainClass(serverLaunchJar);
+
+        return new LaunchData(serverJar, serverLaunchJar, mainClass);
+    }
+
+    private static Properties readProperties() throws IOException {
+        Properties properties = new Properties();
+
+        URL config = getConfigFromResources();
+
+        if (config == null) {
+            throw new RuntimeException("Jar does not contain unattended install.properties file");
+        }
+
+        try (InputStream is = config.openStream()) {
+            properties.load(is);
```

Comment:
This should use the reader based `load()` and a reader with explicit UTF-8

Reply:
I think I took this right from loader, worth checking there.